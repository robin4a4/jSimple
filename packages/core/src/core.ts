const select = <TElement extends HTMLElement>(selector: string) =>
  document.querySelector(selector) as TElement | null;

const selectAll = <TElement extends HTMLElement>(selector: string) =>
  [...document.querySelectorAll(selector)] as Array<TElement> | null;

const create = <TCollection extends HTMLCollection>(htmlString: string) => {
  const template = document.createElement("template");
  template.innerHTML = htmlString.trim();
  return template.content.children as TCollection;
};

function isObjKey<T>(key: PropertyKey, obj: T): key is keyof T {
  return key in obj;
}

type TEachCallback<TObj> = (key: keyof TObj, value: TObj[keyof TObj]) => any;

const each = <TObj>(obj: TObj, callback: TEachCallback<TObj>) => {
  for (const [key, value] of Object.entries(obj)) {
    if (isObjKey(key, obj)) callback(key, value);
  }
};

Array.prototype.first = function () {
  return this[0];
};

Array.prototype.last = function () {
  return this[this.length - 1];
};

NodeList.prototype.first = function () {
  return [...this][0];
};

NodeList.prototype.last = function () {
  return [...this][this.length - 1];
};

HTMLElement.prototype.addClass = function (classes: string) {
  this.classList.add(classes);
  return this;
};

HTMLElement.prototype.removeClass = function (classes: string) {
  this.classList.remove(classes);
  return this;
};

HTMLElement.prototype.toggleClass = function (classes: string) {
  this.classList.toggle(classes);
  return this;
};

HTMLElement.prototype.html = function (html: string) {
  this.innerHTML = html;
  return this;
};

HTMLElement.prototype.empty = function () {
  this.innerHTML = "";
  return this;
};

Element.prototype.on = function (event: string, callback: () => void) {
  this.addEventListener(event, callback);
};

Element.prototype.click = function (callback: () => void) {
  this.on("click", callback);
};

const $ = {
  select,
  selectAll,
  create,
  each,
};

export default $;
