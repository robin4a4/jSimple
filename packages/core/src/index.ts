declare global {
  interface Array<T> {
    first: () => T;
    last: () => T;
  }
  interface NodeList {
    first: () => Node;
    last: () => Node;
  }
  interface HTMLElement {
    addClasses: (classes: string) => HTMLElement;
    removeClasses: (classes: string) => HTMLElement;
    toggleClasses: (classes: string) => HTMLElement;
    html: (html: string) => HTMLElement;
    empty: () => HTMLElement;
    on: (event: string, callback: () => void) => void;
    onClick: (callback: () => void) => void;
  }
}

/**
 * jSimple methods
 */
const select = <TElement extends HTMLElement>(selector: string) =>
  document.querySelector<TElement>(selector);

const selectAll = <TElement extends HTMLElement>(selector: string) => [
  ...document.querySelectorAll<TElement>(selector),
];

const create = <TCollection extends HTMLCollection>(htmlString: string) => {
  const template = document.createElement("template");
  template.innerHTML = htmlString.trim();
  return template.content.children as TCollection;
};

/**
 * reactivity methods
 */
export type TSignal<TSignalValue> = [
  () => TSignalValue,
  (newValue: TSignalValue) => void
];
type TExecuteEffect = () => void;
type TObserver = { execute: TExecuteEffect };

const globalContext: Array<TObserver> = [];

const signal = function <TSignalValue>(
  value: TSignalValue
): TSignal<TSignalValue> {
  const subscriptions = new Set<TObserver>();

  const read = (): TSignalValue => {
    const observer = globalContext[globalContext.length - 1];
    if (observer) subscriptions.add(observer);
    return value;
  };
  const write = (newValue: TSignalValue) => {
    value = newValue;
    for (const sub of [...subscriptions]) {
      sub.execute();
    }
  };
  return [read, write];
};

const effect = function (func: any) {
  const execute: TExecuteEffect = () => {
    globalContext.push(observer);
    try {
      func();
    } finally {
      globalContext.pop();
    }
  };

  const observer: TObserver = {
    execute,
  };

  execute();
};

/**
 * New prototypes methods
 */
Array.prototype.first = function () {
  return this[0];
};

Array.prototype.last = function () {
  return this[this.length - 1];
};

NodeList.prototype.first = function () {
  return [...this][0];
};

NodeList.prototype.last = function () {
  return [...this][this.length - 1];
};

HTMLElement.prototype.addClasses = function (classes: string) {
  classes.split(" ").forEach((className) => {
    this.classList.add(className);
  });
  return this;
};

HTMLElement.prototype.removeClasses = function (classes: string) {
  classes.split(" ").forEach((className) => {
    this.classList.remove(className);
  });
  return this;
};

HTMLElement.prototype.toggleClasses = function (classes: string) {
  classes.split(" ").forEach((className) => {
    this.classList.toggle(className);
  });
  return this;
};

HTMLElement.prototype.html = function (html: string) {
  this.innerHTML = html;
  return this;
};

HTMLElement.prototype.empty = function () {
  this.innerHTML = "";
  return this;
};

HTMLElement.prototype.on = function (
  event: string,
  callback: (ev: Event) => void
) {
  this.addEventListener(event, (ev) => callback(ev));
};

HTMLElement.prototype.onClick = function (callback: () => void) {
  this.on("click", callback);
};

/**
 * Data fetching
 */
enum RequesterMethods {
  Get = "GET",
  Post = "POST",
  Put = "PUT",
  Delete = "DELETE",
}
enum RequesterType {
  Json = "JSON",
  Xml = "XML",
}
interface RequesterOptions {
  type: RequesterType;
  method?: RequesterMethods;
  headers: {
    Accept: string;
    "Content-Type": string;
    Authorization: string;
  };
  body?: string;
}
const jsonResponse = (response: Response) => response.json();
const xmlResponse = (response: Response) => response.text();

async function requester(url: string, options?: RequesterOptions) {
  const results: { content: Promise<any> | null; error: Error | null } = {
    content: null,
    error: null,
  }; // TODO: better type

  let resultFunction = jsonResponse;
  if (options?.type === RequesterType.Xml) resultFunction = xmlResponse;

  const response = await fetch(url, { method: "GET", ...options });
  if (response.ok) {
    results.content = await resultFunction(response);
  } else {
    results.error = new Error("server error");
  }
  return results;
}

function GET(url: string, options: RequesterOptions) {
  return requester(url, { method: RequesterMethods.Get, ...options });
}
function POST(url: string, options: RequesterOptions) {
  return requester(url, { method: RequesterMethods.Post, ...options });
}
function PUT(url: string, options: RequesterOptions) {
  return requester(url, { method: RequesterMethods.Put, ...options });
}
function DELETE(url: string, options: RequesterOptions) {
  return requester(url, { method: RequesterMethods.Delete, ...options });
}

/**
 * jSimple object definition
 */
const $ = {
  select,
  selectAll,
  create,
  signal,
  effect,
  fetch: {
    GET,
    POST,
    PUT,
    DELETE,
  },
};

export default $;
