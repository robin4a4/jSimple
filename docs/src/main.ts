// import { define, signal, callback } from "@jsimple/custom-element";

// @define("fancy-button")
// export class FancyButton extends HTMLElement {
//   @signal("bonjour") nameSignal: any;
//   @signal(true) isOpenSignal: any;
//   isOpen: any;
//   setIsOpen: any;

//   @callback
//   handleClick() {
//     console.log(this.isOpen());
//     this.setIsOpen(!this.isOpen());
//   }

//   connectedCallback() {
//     this.html(`
//     <div>
//     <button type="button" $click="setIsOpen(!isOpen())">{name()}</button>
//     <div $display="isOpen()">lorem ipsum</div>
//     </div>
//     `);
//   }
// }

import $ from "@jsimple/core";

const [isOpen, setIsOpen] = $.signal(false);
const app = $.select("#app");
const btn = $.select("#btn");

$.effect(() => {
  console.log("test");
});

btn?.onClick(() => {
  setIsOpen(!isOpen());
  console.log(isOpen());
});
