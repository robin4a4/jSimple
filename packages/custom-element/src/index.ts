import $, { DOMRender } from "@jsimple/core";

const symbol = Symbol.for("jsimple");

enum Decorator {
  Define = "define",
  Signal = "signal",
  Callback = "callback",
}

function capitalizeFirstLetter(string: string) {
  return string.charAt(0).toUpperCase() + string.slice(1);
}

function formatSignalString(signalString: string) {
  const getterString = signalString.replace("Signal", "");
  const setterString = `set${capitalizeFirstLetter(getterString)}`;
  return [getterString, setterString];
}

function meta(proto: Record<PropertyKey, unknown>, name: string): Set<string> {
  if (!Object.prototype.hasOwnProperty.call(proto, symbol)) {
    const parent = proto[symbol] as Map<string, Set<string>> | undefined;
    const map = (proto[symbol] = new Map<string, Set<string>>());
    if (parent) {
      for (const [key, value] of parent) {
        map.set(key, new Set(value));
      }
    }
  }
  const map = proto[symbol] as Map<string, Set<string>>;
  if (!map.has(name)) map.set(name, new Set<string>());
  return map.get(name)!;
}

export const define = (name: string) => {
  return function (cls: CustomElementConstructor) {
    const connectedCallback = cls.prototype.connectedCallback || function () {};
    cls.prototype.connectedCallback = function () {
      const domSetData = new Set([
        ...meta(cls.prototype, Decorator.Signal),
        ...meta(cls.prototype, Decorator.Callback),
      ]);
      const domObjectData: any = {};
      for (let propertyString of domSetData) {
        domObjectData[propertyString] = cls.prototype[propertyString];
      }
      connectedCallback.call(this);
      DOMRender(domObjectData, this);
    };

    window.customElements.define(name, cls);
  };
};

export function signal<K extends string>(
  proto: Record<K, unknown>,
  key: K
): any {
  const [getterString, setterString] = formatSignalString(key);
  const signal = $.signal(false);
  Object.defineProperty(proto, key, {
    configurable: true,
    value: signal,
  });
  Object.defineProperty(proto, getterString, {
    configurable: true,
    value: signal[0],
  });
  Object.defineProperty(proto, setterString, {
    configurable: true,
    value: signal[1],
  });
  meta(proto, Decorator.Signal).add(getterString);
  meta(proto, Decorator.Signal).add(setterString);
}

export function callback(target: any, propertyKey: string) {
  meta(target, Decorator.Callback).add(propertyKey);
}
