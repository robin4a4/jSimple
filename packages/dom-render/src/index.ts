// @ts-ignore
import { tmpl } from "riot-tmpl";
import $ from "@jsimple/core";

/**
 * Reactive dom attribute
 *
 * TODO: - manage open() native function and others
 *       - remove tmpl dependency
 */
interface HTMLElementWithData extends HTMLElement {
  data?: string;
}

export function DOMRender<TContext>(
  context: TContext,
  mount: HTMLElementWithData
) {
  let current;
  let next = mount.firstChild as HTMLElementWithData;
  while (next) {
    current = next;
    const type = current.nodeType;
    if (type === 1 && current.hasAttribute("data-define")) break;

    processNode<TContext>(current, context);
    next = DOMRender(context, current) || next.nextSibling;
  }
  return next;
}

function processNode<TContext>(el: HTMLElementWithData, context: TContext) {
  const type = el.nodeType;

  // element
  if (type === 1) {
    for (const { name, value } of [...el.attributes]) {
      if (name[0] === "$") {
        const jAttr = name.slice(1);
        const eventList = [
          "click",
          "mouseover",
          "mouseout",
          "change",
          "keydown",
        ];
        if (eventList.find((eventName) => eventName === jAttr)) {
          el.addEventListener(jAttr, () => tmpl(`{${value}}`, context));
        }
        if (jAttr === "display") {
          $.effect(() => {
            el.style.display = tmpl(`{${value}}`, context) ? "block" : "none";
          });
        }
      }
    }
  }
  // text node
  else if (type === 3) {
    if (tmpl.hasExpr(el.data) && el.data) {
      if (/(\(.*\))?(\.)?/g.test(el.data)) {
        $.effect(() => {
          const expr = tmpl(el.data, context);
          if (expr) {
            el.data = expr;
          }
        });
      } else el.data = tmpl(el.data, context);
    }
  }
}

export function run<TContext>(funcArray: Array<() => TContext>) {
  funcArray.forEach((func) => {
    const mountEl = $.select(
      `[data-define='${func.name
        .replace(/([a-z0â€“9])([A-Z])/g, "$1-$2")
        .toLowerCase()}']`
    );
    if (!mountEl) return;
    DOMRender<TContext>(func(), mountEl);
  });
}
