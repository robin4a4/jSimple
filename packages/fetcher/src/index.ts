import $, { TSignal } from "@jsimple/core";

/**
 * Data fetching
 */
enum RequesterMethods {
  Get = "GET",
  Post = "POST",
  Put = "PUT",
  Delete = "DELETE",
}
enum RequesterType {
  Json = "JSON",
  Xml = "XML",
}
interface RequesterOptions {
  type?: RequesterType;
  method: RequesterMethods;
  headers?: {
    Accept: string;
    "Content-Type": string;
    Authorization: string;
  };
  body?: URLSearchParams | FormData | Blob | BufferSource;
}

type CRUDOptionType = Omit<RequesterOptions, "method">;

const jsonResponse = (response: Response) => response.json();
const xmlResponse = (response: Response) => response.text();

async function requester<TData>(url: string, options: RequesterOptions) {
  let resultFunction = jsonResponse;
  if (options?.type === RequesterType.Xml) resultFunction = xmlResponse;

  const response = await fetch(url, options);
  return resultFunction(response) as Promise<TData>;
}

export function GET<TData>(
  url: string,
  options?: Omit<CRUDOptionType, "body">
) {
  return requester<TData>(url, { method: RequesterMethods.Get, ...options });
}
export function POST<TData>(url: string, options?: CRUDOptionType) {
  return requester<TData>(url, { method: RequesterMethods.Post, ...options });
}
export function PUT<TData>(url: string, options?: CRUDOptionType) {
  return requester<TData>(url, { method: RequesterMethods.Put, ...options });
}
export function DELETE<TData>(url: string, options?: CRUDOptionType) {
  return requester<TData>(url, { method: RequesterMethods.Delete, ...options });
}

/**
 * Query management (like a super tiny tanstack-query)
 * @param promise
 * @param key
 * @returns
 */
export const load = <TData>(
  promise: Promise<TData>,
  key: Array<string | TSignal<any>>
) => {
  const [data, setData] = $.signal<TData | null>(null);
  const [isLoading, setIsLoading] = $.signal(true);
  const [isError, setIsError] = $.signal(false);
  const [error, setError] = $.signal(null);
  const stringifiedKey = JSON.stringify(key);
  const cachedData = sessionStorage.getItem(stringifiedKey);
  if (!cachedData) {
    promise
      .then((data) => {
        setData(data);
        sessionStorage.setItem(stringifiedKey, JSON.stringify(data));
      })
      .catch((err) => {
        setIsError(true);
        setError(err);
      })
      .finally(() => setIsLoading(false));
  } else {
    setData(JSON.parse(cachedData));
    setIsLoading(false);
  }

  return {
    data,
    isLoading,
    isError,
    error,
  };
};
