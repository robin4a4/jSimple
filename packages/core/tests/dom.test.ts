// @vitest-environment jsdom
import { expect, test } from "vitest";
import $ from "../src";

function domExample() {
  const div = document.createElement("div");
  div.innerHTML = `
    <label for="username">Username</label>
    <input id="username" />
    <button>Print Username</button>
  `;
  return div;
}

function eventExample() {
  const div = domExample();
  const button = div.querySelector("button");
  const input = div.querySelector("input");
  button?.addEventListener("click", () => {
    // let's pretend this is making a server request, so it's async
    // (you'd want to mock this imaginary request in your unit tests)...
    setTimeout(() => {
      const printedUsernameContainer = document.createElement("div");
      printedUsernameContainer.innerHTML = `
        <div data-testid="printed-username">${input?.value}</div>
      `;
      div.appendChild(printedUsernameContainer);
    }, Math.floor(Math.random() * 200));
  });
}

test.skip("select", () => {
  const div = domExample();
  expect($.select(".div")).toBe(div);
});

test.skip("selectAll", () => {
  const div = domExample();
  const div2 = domExample();
  expect($.selectAll(".div")).toBe([div, div2]);
});

const array = [1, 2];

test("first", () => {
  expect(array.first()).toBe(1);
});

test("last", () => {
  expect(array.last()).toBe(2);
});

test("add one class", () => {
  const div = domExample();
  div.addClasses("flex");
  expect(div.getAttribute("class")).toBe("flex");
});

test("add multiple classes and remove", () => {
  const div = domExample();
  div.addClasses("flex border");
  expect(div.getAttribute("class")).toBe("flex border");
  div.removeClasses("flex border");
  expect(div.getAttribute("class")).toBe("");
});

test("toggle class", () => {
  const div = domExample();
  div.toggleClasses("flex");
  expect(div.getAttribute("class")).toBe("flex");
  div.toggleClasses("flex");
  expect(div.getAttribute("class")).toBe("");
});
