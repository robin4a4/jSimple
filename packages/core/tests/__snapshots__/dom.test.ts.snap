// Vitest Snapshot v1

exports[`examples of some things 1`] = `
<div>
  
    
  <label
    for="username"
  >
    Username
  </label>
  
    
  <input
    id="username"
  />
  
    
  <button>
    Print Username
  </button>
  
  
  <div>
    
        
    <div
      data-testid="printed-username"
    >
      Ada Lovelace
    </div>
    
      
  </div>
</div>
`;
